### PR TITLE
Validate input CSV files when reading.

### PR DESCRIPTION
**Pull request by lkitching**
Issue #37 - Create a CSV parser which validates and transforms data
rows according to a declarative specification. Use this parser when
parsing input files for the codelist and components pipelines as well
as the columns configuration file within the cube pipeline. This
parser checks that all required columns are present in the input and
no unknown columns are specified.

Rename csv/read-csv-records to read-csv-maps and name the new parser
read-csv-records.

### CODE REVIEW

**Comment by RickMoynihan** [comment]
Out of interest @lkitching what do the validation errors look like?

**Comment by lkitching** [comment]
@RickMoynihan - Cell validation errors have the form `Invalid cell in column [column name] row [row number]: [message]`. There are other message formats for invalid columns in the header.

File: src/table2qb/configuration/column.clj [hunk 1]
```diff
@@ -54,60 +56,67 @@
...
+                   :key :value_template
+                   :transform (csv/optional uri-template)}
+                  {:title "datatype"
+                   :key :datatype
+                   :transform (csv/optional validate-csvw-datatype)}
+                  {:title "value_transformation"
+                   :key :value_transformation
+                   :transform (csv/optional (csv/validate-mapping column-transformers))}])
+
+(defn normalise-column-record [row]
```
Comment (Robsteranium) [inline src/table2qb/configuration/column.clj:112]: I'm a little wary of using the term "normalise". This makes me think of structural changes - e.g. un-pivoting to 3rd Normal Form.

Is this operation something more like `remove-optional-columns`?

File: test/table2qb/configuration/column_test.clj [hunk 2]
```diff
@@ -2,58 +2,38 @@
   (:require [clojure.test :refer :all]
             [table2qb.configuration.column :refer :all]))
```
Comment (Robsteranium) [inline test/table2qb/configuration/column_test.clj:4]: I wonder if it makes sense to add some examples to `columns-test`? We have unit tests in `column-test` for each validation, and `csv-test` checks that the parser works but we're not checking that we've specified the columns input correctly.

File: src/table2qb/configuration/column.clj [hunk 3]
```diff
@@ -54,60 +56,67 @@
...
-      (when vt
-        {:value_transformation (resolve-value-transformation vt)}))))
+(def ^:private attachment->type {"qb:dimension" :dimension
+                                 "qb:measure" :measure
+                                 "qb:attribute" :attribute})
+
+(defn validate-column-type [row column value]
+  (if (string/blank? value)
+    :value
+    (let [type (get attachment->type (string/trim value) ::missing)]
```
Comment (Robsteranium) [inline src/table2qb/configuration/column.clj:66]: As per [this comment](https://github.com/Swirrl/table2qb/pull/115#issuecomment-616500121), perhaps we should be stricter here and not trim white space from the cell values?

Comment (lkitching) [reply src/table2qb/configuration/column.clj:66]: It looks like we can just use `csv/validate-mapping` here and update `cell-enum-validation-message` to include the quoted value e.g. `Invalid value "$value", expected one...'?

File: test/table2qb/configuration/column_test.clj (hunk 2 above)
Comment (lkitching) [reply test/table2qb/configuration/column_test.clj:4]: Yes that would be useful!

Comment (Robsteranium) [reply test/table2qb/configuration/column_test.clj:4]: Done in 689660c.

File: src/table2qb/configuration/column.clj (hunk 3 above)
Comment (Robsteranium) [reply src/table2qb/configuration/column.clj:66]: I've removed the `string/trim` in e3e280b and the existing validation message already covers this well!

**Pull request merged**